api/odilo.py

```diff
@@ -0,0 +1,532 @@
+        if response.status_code == 200:
+            return self.hold_from_odilo_hold(licensepool.collection, data)
+
+        if response.status_code in (403, 404):
+            if not data or 'errors' not in data or len(data['errors']) <= 0:
+                raise CannotHold()
+
+            error = data['errors'][0]
+            code = error['id']
+            description = ''
+            if 'description' in error:
+                description = error['description']
+
+            if code == 'LOAN_ALREADY_RESERVED':
+                raise AlreadyOnHold(description)
+            elif code == 'ERROR_DATA_NOT_FOUND':
+                raise NotFoundOnRemote(code + ', record: ' + record_id)
+            elif code == 'PatronExceededHoldLimit':
+                raise PatronHoldLimitReached(description)
+
+        raise CannotHold()
+
+    def release_hold(self, patron, pin, licensepool):
+        """Release a patron's hold on a book.
+        """
+        hold_id = licensepool.identifier.identifier
+        url = self.RELEASE_HOLD_ENDPOINT.replace("{holdId}", hold_id)
+        payload = json.dumps(dict(patronId=patron))
+
+        response = self.patron_request(patron, pin, url, extra_headers={}, data=payload, method='POST')
+
+        if response.status_code == 200:
+            return True
+
+        if not response.content:
+            raise CannotReleaseHold()
+
+        data = response.json()
+        if 'errors' not in data or len(data['errors']) <= 0:
+            raise CannotReleaseHold()
+
+        error = data['errors'][0]
+        code = error['id']
+        description = error['description']
+
+        if code == 'HOLD_NOT_FOUND':
+            return True
+
+        raise CannotReleaseHold(description)
+
+
+class OdiloCirculationMonitor(CollectionMonitor):
+    """Maintain LicensePools for recently changed Odilo titles
+    """
+    SERVICE_NAME = "Odilo Circulation Monitor"
+    INTERVAL_SECONDS = 500
+    PROTOCOL = ExternalIntegration.ODILO
+
+    def __init__(self, _db, collection, api_class=OdiloAPI):
+        """Constructor."""
+        super(OdiloCirculationMonitor, self).__init__(_db, collection)
+        self.api = api_class(_db, collection)
+
+    def run_once(self, start, cutoff):
+        self.log.info("Starting recently_changed_ids, start: " + str(start) + ", cutoff: " + str(cutoff))
+
+        start_time = datetime.datetime.now()
+        self.recently_changed_ids(start)
+        finish_time = datetime.datetime.now()
+
+        time_elapsed = finish_time - start_time
+        self.log.info("recently_changed_ids finished in: " + str(time_elapsed))
+
+    def recently_changed_ids(self, start):
+        modification_date = None
+        if start:
+            if isinstance(start, datetime.date):
+                modification_date = start.strftime('%Y-%m-%d')  # Format YYYY-MM-DD
+            elif isinstance(start, basestring):
+                modification_date = start
+
+        self.all_ids(modification_date)
```

Comment:
You can just call `all_ids` and let it parse the modification date.

Reply:
Ok, I've made what you say first.
Related to the format of the modification date, it is enough for us to use just the date and not the time.

Reply:
Does the API show the modification date but not the time?

I'm concerned that every time this runs over the course of a day, it's going to pick up more and more books, until the stroke of midnight, at which point it will pick up zero books and the process will repeat. This shouldn't hurt the integrity of the data--like I said, I'm just concerned about performance.

Reply:
We use just the modification date and not the time due to internal restrictions of the DB and the app.